src/kernel/src/fs/file.rs

```diff
@@ -18,19 +13,28 @@ use thiserror::Error;
 /// An implementation of `file` structure.
 #[derive(Debug)]
 pub struct VFile {
-    ty: VFileType,      // f_type
-    flags: VFileFlags,  // f_flag
-    offset: Gutex<i64>, // f_offset
+    ty: VFileType,             // f_type
+    flags: VFileFlags,         // f_flag
+    vnode: Option<Arc<Vnode>>, // f_vnode
+    offset: Gutex<u64>,        // f_offset
+    backend: Box<dyn FileBackend>,
```

Comment:
I'm not sure I fully understand the decision to use dyn here. I understand it for cases where FreeBSD has just some ops + void * data, but not here, where FreeBSD has ops + data + type. Here, we technically allow inconsistent state between the VFileType and backend. I get that using enums is a bit tedious (there's a lot of matches), but we could still utilize your new VnodeFileBackend CDevFilebackend in the enum as it was before.

Reply:
Actually on FreeBSD the `ops` + `data` is decoupled from `type`. An example is `devfs_ops_f` that use `DTYPE_VNODE` as a type.

Reply:
Also, dyn is slower that enums (I'm pretty sure) which is why I don't want to use it for no good reason, which seems to me, is the case here.

Reply:
> Actually on FreeBSD the `ops` + `data` is decoupled from `type`. An example is `devfs_ops_f` that use `DTYPE_VNODE` as a type.

I completely missed that :D I guess it makes sense then

Reply:
The dynamic dispatch does not have performance impact in most cases. Unless we can prove the performance problem causes by a dynamic dispatch we don't need to worry about it.

Reply:
Not sure if you know about this. The dynamic dispatch on trait object is just a call on a function pointer, the same as what FreeBSD use extensively on their `ops` things. That mean in worst case our performance will be on par with FreeBSD.

Reply:
I do know that. I also did know that generally, the performace impact is negligible. My point was simply that I saw no compelling reason to refactor to use dyn, but you gave me one. I automatically assumed that DTYPE_DEV is used for devfs_ops_f, but it seems that it's actually only used for some Linux compatibility thing. Interesting.